src/primitives/transaction.h

```diff
@@ -197,6 +189,13 @@ class CTxOut
 
 struct CMutableTransaction;
 
+struct TransactionSerParams {
+    const bool allow_witness;
+    SER_PARAMS_OPFUNC
+};
+static constexpr TransactionSerParams TX_WITH_WITNESS{.allow_witness = true};
```

Comment:
The approach we've got looks like:

```c++
    dumb_serializer << SER_PARAMS(object)
    // file, CTransaction
```

which then gets translated to:

```c++
    dumb_serializer << smart_object
    // file, ParamsWrapper<TransactionSerParams, CTransaction>
```

then

```c++
    smart_serializer.Serialize(object, SER_PARAMS)
    // ParamsStream<TransactionSerParams, dumb_serializer>, CTransaction
```

The advantage of saying `TX_WITH_WITNESS` everywhere is that way we're being explicit about how objects are serialised to dumb streams everywhere, so we get compile errors if we're ever potentially inconsistent.

But I don't think that makes sense for multiprocess: we don't need to be using dumb streams in the first place there because the whole point is to pass objects between our own processes, no? So afaics, we should be using a smart serializer from the start; ie one that provides a `GetParams()` that returns something that can be be converted to `TransactionSerParams`. So doing `ParamsStream ss{TX_WITH_WITNESS, stream}` in `CustomBuildField` and `CustomReadField` would be an easy way to to start?

If you want to pass around both `CNetAddr` and `CAddress` objects as well, I think this would work:

```c++
// serialize.h
template<typename... SerParams>
class MultiSerParam
{
private:
    const std::tuple<SerParams...> m_params;
public:
    constexpr MultiSerParam(const SerParams&... t) : m_params{t...} { }

    template<typename T>
    constexpr operator const T&() const { return std::get<T>(m_params); }
};

// ipc/capnp/common-types.h ?
static constexpr MultiSerParam INTERNAL_SER_PARAMS(
  TX_WITH_WITNESS,
  CNetAddr::V2,
  CAddress::V2_NETWORK
);

void f(const CNetAddr& netaddr, const CAddress& addr, const CTransaction& tx)
{
    DataStream substream;
    ParamsStream ss{INTERNAL_SER_PARAMS, substream};
    ss << netaddr;
    ss << tx;
    ss << addr;
}
```

Note that making this actuall compile also requires tweaking `CNetAddr` like so:

```c++
-        if (s.GetParams().enc == Encoding::V2) {
+        CNetAddr::SerParams p(s.GetParams());
+        if (p.enc == Encoding::V2) {
```

Reply:
Thanks! That approach of being able to combine different params is clever and solves my problem, avoiding boilerplate from having to create different streams for each object type.

Maybe at some point it would be useful to support default parameter values, too, so parameters like CNetAddr::V2 and CAddress::V2_NETWORK could be assumed if not specified. For some serialization parameters it may be better to be explicit, but for others there could be downsides. For example if I hardcode V2 in multiprocess code and a V3 serialization is added later it could cause subtle bugs if V3 data is silently lost in IPC calls.

Reply:
If the issue is the separation between the `CNetAddr::SerParams` declaration and the `MultiSerParams(..,CNetAddr::V2,..)` use, could instead make it:

```c++
class MultiSerParam
{
public:
    template<typename T>
    operator T&() const { return T::GetIPCSerParams(); }
};
constexpr MultiSerParam INTERNAL_SER_PARAMS;

class CNetAddr {
    struct SerParams {
        const Encoding enc;
        static SerParams GetIPCSerParams() { return {Encoding::V2}; }
        SER_PARAMS_OPFUNC
   };
};
```

That leaves it explicit for the dumb streams, and provides a clear, localised default for internal serialization of the object?

(I'm more worried about changing a default from v2 to v3 causing subtle bugs when deserialising network data or from a file, so I prefer being explicit everywhere and finding some other way to fix the other annoyances)

Reply:
I guess the idea of having default parameters is off the table then. So probably the way forward to make IPC code safer is have an IPC serialization parameters like your suggestion.

For now though I left the safety issue aside and implemented a way for IPC code to pass multiple stream parameters to ParamsStream in #28929 (used in #10102 [here](https://github.com/ryanofsky/bitcoin/blob/7ba123f1b47a4a790370c9a5975b1e0b4026a241/src/ipc/capnp/common-types.h#L52-L56)), so the same stream can be used for all serialization, and there is no need for code that creates different streams depending on the object type. This solves the problem of parameters increasing boilerplate in multiprocess code.

Reply:
I'm just saying what my opinion is; doesn't mean I'm necessarily right. What you've come up with looks pretty good to me though.